hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/stats/SidecarInstrumentationImpl.java

```diff
@@ -0,0 +1,208 @@
+import java.time.Instant;
+import java.util.EnumMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Holds the accumulators needed for the metrics/logging of a sidecar record (corresponding, most
+ * likely, to a record block. There's a separate accumulator for each kind of sidecar record
+ * (bytecode, state change, actions).
+ */
+public record SidecarInstrumentationImpl(
+        @NonNull EnumMap<SidecarType, SidecarAccumulators> accumulators,
+        @NonNull Clock clock,
+        @NonNull AtomicInteger copies)
+        implements SidecarInstrumentation {
+
+    /** Constructs the sidecar instrumentation */
+    public SidecarInstrumentationImpl() {
+        this(new EnumMap<>(SidecarType.class), Clock.systemUTC(), new AtomicInteger());
+        createAccumulators();
+    }
+
+    /** Constructs the sidecar instrumentation, allowing the injection of a `Clock` (for testing) */
+    public SidecarInstrumentationImpl(@NonNull Clock clock) {
+        this(new EnumMap<>(SidecarType.class), clock, new AtomicInteger());
+        createAccumulators();
+    }
+
+    /** Create an accumulator for each kind of sidecar record */
+    private void createAccumulators() {
+        for (final var e : SidecarType.values())
+            accumulators.put(e, new SidecarAccumulators(e.name()));
+    }
+
+    /**
+     * Add a sample of the serialized size of a sidecar transaction, for a particular kind of
+     * sidecar
+     */
+    @Override
+    public void addSample(@NonNull final SidecarType type, long size) {
+        final var acc = accumulators.get(type);
+        acc.serializedSize().addSample(size);
+    }
+
+    /**
+     * Add a sample of the compute duration of a sidecar transaction, for a particular kind of
+     * sidecar
+     */
+    @Override
+    public void addSample(@NonNull final SidecarType type, @NonNull final Duration duration) {
+        final var acc = accumulators.get(type);
+        acc.computeDuration().addSample(duration.toNanos());
+    }
+
+    /**
+     * Add a sample of both serialized size and compute time for a sidecar transaction, for a
+     * particular kind of sidecar
+     */
+    @Override
+    public void addSample(
+            @NonNull final SidecarType type, long size, @NonNull final Duration duration) {
+        final var acc = accumulators.get(type);
+        acc.serializedSize().addSample(size);
+        acc.computeDuration().addSample(duration.toNanos());
+    }
+
+    /**
+     * Add the entire accumulated values of a SidecarInstrumentation to this one (which is a summary
+     * of multiple records
+     */
+    @SuppressWarnings("java:S2445")
+    // sync on method parameter _OK_ because adding samples goes only in _one_ direction ever (from
+    // more specific accumulator to more general accumulator)
+    @Override
+    public void addSamples(@NonNull final SidecarInstrumentation samples) {
+        if (samples instanceof SidecarInstrumentationImpl impl) {
+            synchronized (this) {
```

Comment:
This looks scary. 😄 Is it possible to check the precondition to make sure no dead lock occurs?

Reply:
Yes.  The synchronization is to do an _atomic_ rollup from multiple metrics (each of which _itself_ is accumulated atomically of course).  I myself doubt there'd ever be any contention.  Let's see if I can _prove_ there can't be any contention ... more coming ...

Reply:
oh by the way it can't be either _deadlock_ or _livelock_ because the rollups go only _one way_ - from more specific/partial to more general/complete metrics.  That is, we compute metrics on an individual tx then roll-up to metrics on an entire sidecar block/file.  So there's never a case where you'll addSample in the other direction.  Admittedly, this can't be enforced in this API.  Makes me wonder if I can design an API where it _is_ enforced?  Typically you'd do that by having the more specific hold a handle to the more general (or vice versa, but only one or the other) but in this case the individual specific per-tx metrics are created far far away from where the sidecar blocks are created and accumulate them, and they only come in over time (via a queue).  So I'm not sure what it would look like.  Or if it is worth it.  I extended the usage comment at the method decl in the interface.